apkid/rules/elf/obfuscators.yara

```diff
@@ -130,7 +131,87 @@ rule ollvm_v6_0 : obfuscator
     $based_on      = "(based on Obfuscator-LLVM 6.0."
 
   condition:
-    all of them and not ollvm_v6_0_strenc
+    is_elf and
+    all of them and
+    not ollvm_v6_0_strenc
+}
+
+rule ollvm_v9_strenc : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM version 9.x (string encryption)"
+    sample      = "2314ec0053d829d424a82f702188fcb525cefce4feeef096f0855339b897a5d1"
+    author      = "Eduardo Novella"
+
+  strings:
+    $clang_version = /clang version \d\.\d\.\d /
+    $strenc        = /\.datadiv_decode[0-9]{18,20}/
+    $ollvm         = "(based on Obfuscator-LLVM 9."
+
+  condition:
+    is_elf and all of them
+}
+
+rule ollvm_tll : obfuscator
+{
+  meta:
+    description = "Obfuscator-LLVM TLL (string encryption)"
+    url         = "https://github.com/yazhiwang/ollvm-tll.git"
+    sample      = "1f010330e9ac90f00d11aa37fdca25c437ad6f4b1302f6d7aa48b91ef22cc107"
+    author      = "Eduardo Novella"
+
+  strings:
+    /**
+      .datadiv_decode7760209850571766755
+      Android clang version 5.0.300080  (based on LLVM 5.0.300080)
+      clang version 6.0.0 (tags/RELEASE_600/final) (https://github.com/yazhiwang/ollvm-tll.git a38559e4c13359073102793c0a734bb1add3d5ff)
+    */
+    $clang_version = /clang version \d\.\d\.\d \(tags\/RELEASE\_\d{3}\/final\)/
+    $strenc        = /\.datadiv_decode[0-9]{18,20}/
+    $url           = "https://github.com/yazhiwang/ollvm-tll.git"
+
+  condition:
+    is_elf and all of them
+}
+
+rule ollvm_armariris : obfuscator
+{
+  meta:
+    description = "Armariris Obfuscator-LLVM (string encryption)"
+    url         = "https://github.com/GoSSIP-SJTU/Armariris"
+    sample      = "d22c2f53bab6fa2ab7bdb4f7acabb419e3ee3163bb758c4f7a013d07a8b09e12" // aka malware Joker
+    author      = "Eduardo Novella"
+
+  strings:
+    // clang version 3.9.1 (tags/RELEASE_391/final)
+    // clang version 5.0.1 (tags/RELEASE_501/final)
+    // .datadiv_decode14660921177804423408
+    $clang_version = /clang version \d\.\d\.\d \(tags\/RELEASE\_\d{3}\/final\)/
+    $strenc        = /\.datadiv_decode[0-9]{18,20}/
+
+  condition:
+    is_elf and
+    all of them and
+    not ollvm_tll and
+    for any i in (0..elf.symtab_entries): (elf.symtab[i].name contains ".datadiv_decode")
```

Comment:
I really like checking the symtab_entries as it's way more specific than just looking for a string. Why contains here and not matching the regex?

Reply:
Do you mean using the keyword `matching`? Don't remember well why this keyword didn't want to work out. The Yara unittests had this loop in one of the tests. Did it work like a charm for my case, because regex + matching was not compiling. As we cannot use regex within the loop, the only option I came with was using the keyword `contains`.

Reply:
`and for any i in (0..elf.symtab_entries): (elf.symtab[i].name matches /\.datadiv_decode[\d]{18,20}/)` compiles. Did you try this?

Reply:
Compiled! Good improvement & changes were made in several locations. Ready to be merged

Reply:
what about shorting it a bit using the variable `$strenc`?

```sh
(elf.symtab[i].name matches $strenc)
```